### DISCUSSION TITLE
Validation of QuestionnaireResponse based on Questionnaire

### DISCUSSION THREAD

**Discussion by mmalvik** [FirelyTeam/firely-net-sdk#1905] [Q&A] [answered]
Hi 👋 

Disclaimer: I'm pretty new to FHIR so please bear with me.

I am wondering how you would approach validating a `QuestionnaireResponse`?
For simplicites sake say you have a `Questionnaire` with one question that has three answer options:

```c#
new ItemComponent
{
    LinkId = "myLinkId",
    Required = true,
    Repeats = false,
    Type = QuestionnaireItemType.Choice,
    AnswerOption = new List<AnswerOptionComponent>
    {
        new AnswerOptionComponent
        {
            Value = new FhirString("Yes")
        },
        new AnswerOptionComponent
        {
            Value = new FhirString("No")
        },
        new AnswerOptionComponent
        {
            Value = new FhirString("Do not know")
        }
    }
}
```

How would you go about validating a `QuestionnaireResponse` for this `Questionnaire`?
Is there any built in validation you can use?

**Comment by marcovisserFurore** [comment, marked as answer]
Hi @mmalvik,

Welcome to the world of Fhir!

I think you mean here the semantic validation of a `QuestionnaireResponse` to `Questionnaire`. So the following would be not valid:

```c#
    var item = new QuestionnaireResponse.ItemComponent();
    item.LinkId = "myLinkId";
    item.Answer.Add(new QuestionnaireResponse.AnswerComponent() { Value = new FhirBoolean(true) });
```
because the type of the answer should be a string and not a boolean. 

Unfortunately, our SDK is not able to validate this. The Validator, that comes with the SDK, validates Fhir resources only on their structure, cardinality of all properties, invariants, etc. Not whether a `QuestionnaireResponse` fits with the corresponding `Questionnaire`.

We can add this feature to our roadmap.

**Reply by mmalvik** [reply]
Yes, you undestood my question perfectly.
In an ideal world I could have serialized a `QuestionnaireResponse` and then validated it against a given `Questionnaire`.

Very basic pseudo code to demonstrate my thinking:
```c#
var questionnaireResponse = GetQuestionnaireResponse(); // Get the response and serialize it somehow
var questionnaire = GetQuestionnaire();

var validationResult = ValidateQuestionnaireResponse(questionnaire, questionnaireResponse);
```
I don't know if my thinking is completely correct here though.

Thank you for a quick answer @marcovisserFurore 👍
